### PR TITLE
feat: add tidb vector store

### PR DESCRIPTION
**Pull request by Mini256**

### CODE REVIEW

File: packages/core/src/storage/vectorStore/TiDBVectorStore.ts [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+        // All good?  Keep the connection reference
+        this.db = db;
+      } catch (err: any) {
+        console.error(err);
+        return Promise.reject(err);
+      }
+    }
+
+    return Promise.resolve(this.db);
```
Comment (marcusschiesser) [inline packages/core/src/storage/vectorStore/TiDBVectorStore.ts:108]:
```suggestion
        throw err;
      }
    }

    return this.db;
```

File: packages/core/src/storage/vectorStore/TiDBVectorStore.ts [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    namespace?: string;
+    tableName?: string;
+    dimensions?: number;
+    poolOptions?: PoolOptions;
+    client?: mysql.Pool;
+  }) {
+    this.tableName = config?.tableName ?? TIDB_VECTOR_TABLE;
+    this.namespace = config?.namespace ?? "";
+    this.dimensions = config?.dimensions ?? 1536;
+    this.poolOptions = config?.poolOptions ?? {};
```
Comment (marcusschiesser) [inline packages/core/src/storage/vectorStore/TiDBVectorStore.ts:55]: you can use `Object.assign` to copy the config to this (then you don't need to duplicate your default values)

File: packages/core/src/storage/vectorStore/TiDBVectorStore.ts [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+   */
+  async delete(refDocId: string, deleteKwargs?: any): Promise<void> {
+    const namespaceCriteria = this.namespace.length ? "AND namespace = $2" : "";
+    const sql: string = `DELETE FROM ${this.tableName} WHERE id = $1 ${namespaceCriteria}`;
+    const db = await this.getDb();
+    const params = this.namespace.length
+      ? [refDocId, this.namespace]
+      : [refDocId];
+    await db.query(sql, params);
+    return Promise.resolve();
```
Comment (marcusschiesser) [inline packages/core/src/storage/vectorStore/TiDBVectorStore.ts:227]: no-op

File: packages/core/src/storage/vectorStore/TiDBVectorStore.ts [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+  document: string;
+  metadata: Metadata;
+  embeddings: number[];
+  score: number;
+}
+
+/**
+ * Provides support for writing and querying vector data in TiDB.
+ */
+export class TiDBVectorStore implements VectorStore {
```
Comment (marcusschiesser) [inline packages/core/src/storage/vectorStore/TiDBVectorStore.ts:27]: did you check if the data generated is compatible with the python version (see https://github.com/run-llama/llama_index/blob/main/llama-index-integrations/vector_stores/llama-index-vector-stores-tidbvector/llama_index/vector_stores/tidbvector/base.py) ?

File: packages/core/src/storage/vectorStore/TiDBVectorStore.ts [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    return Promise.resolve();
+  }
+
+  /**
+   * Query the vector store for the closest matching data to the query embeddings
+   * @param query The VectorStoreQuery to be used
+   * @param options Required by VectorStore interface.  Currently ignored.
+   * @returns Zero or more Document instances with data from the vector store.
+   */
+  async query(
```
Comment (marcusschiesser) [inline packages/core/src/storage/vectorStore/TiDBVectorStore.ts:236]: We're using metadataDictToNode and nodeToMetadata from "./utils.js" to de/serialize the node in the metadata. You can reference the other vector store implementations to have a look into this.

**Review by marcusschiesser — COMMENTED** [review]
@Mini256 looking great! Thanks for your contribution. Please have a look at my comments and if you have time it would be great if you could add an example to the `/examples` folder

File: packages/core/src/storage/vectorStore/TiDBVectorStore.ts (hunk 4 above)
Comment (Mini256) [reply packages/core/src/storage/vectorStore/TiDBVectorStore.ts:27]: grest, I will make some adjustments to be consistent with the Python version.

**Review by himself65 — COMMENTED** [review]
any update on this?

**Comment by himself65** [comment]
Closing as stale

**Pull request closed without merging**